File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> mystack = new Stack();
        
        for(int i=0;i<s.length();i++){
            char c = s.charAt(i);
            if (c == '(' || c == '{' || c == '['){
                mystack.push(c);
            }
            else {
                if (mystack.isEmpty())
                    {
                        return false;
                    }
                char top = mystack.pop();
                if ( (c==')' && top != '(')||
                     (c==']' && top != '[')||
                     (c=='}' && top != '{')){
                    return false;
                }
            }
        }
            return mystack.isEmpty();
    }
}

File: 0020-valid-parentheses/0020-valid-parentheses.py
# Time O(n); Space: O(n)
# Just add into the Stack if its a open braces and check if the next character is equal to the top of the stack (else condition)
class Solution:
    def isValid(self, s: str) -> bool:
        mystack = []
        
        for i in s:
            if i == '{' or i == '(' or i == '[':
                if i == '{':
                    mystack.append('}')
                elif i == '(':
                    mystack.append(')')
                elif i == '[':
                    mystack.append(']')
            else:
                # Edge case if it starts with closed braces then stack will be empty so return False
                if not mystack or i != mystack[-1]:
                    return False
                else:
                    mystack.pop()
        # Make sure to return true only of if the stack is empty
        return len(mystack) == 0
                
        
        